server/watchdog/watchdog.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchdog
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	tickMs = 100
+)
+
+var (
+	wdInstance     *watchdog
+	nextActivityId uint64
+)
+
+type activity struct {
+	id              uint64
+	name            string
+	inactiveElapsed int
+}
+
+type watchdog struct {
+	lg *zap.Logger
+
+	ticker              *time.Ticker
+	inactiveTimeoutTick int
+
+	mu         sync.Mutex
+	activities map[uint64]*activity
+
+	stopC   <-chan struct{}
+	cleanup func()
+}
+
+func (wd *watchdog) run() {
+	inactiveTimeout := time.Duration(wd.inactiveTimeoutTick*tickMs) * time.Millisecond
+	wd.lg.Info("Watchdog is running", zap.Duration("inactiveTimeout", inactiveTimeout))
+	for {
+		select {
+		case <-wd.ticker.C:
+			wd.mu.Lock()
+			for _, v := range wd.activities {
+				v.inactiveElapsed++
+				if v.inactiveElapsed > wd.inactiveTimeoutTick/2 {
```

Comment:
The fundamental difference between this design and watchdog pattern in general is that you propose to monitor a specific registered activities. 

I assumed the goal is to monitor watchdog as a whole thing.... 

Watchdog is unhealthy when for the whole 'timeout' it got no 'Start' nor 'Stop/Done' notifications. 
Any such notifications does reset the watchdog. Lack of new activities is especially an alart. 

We do register specific activities only to: 
  a) have actionable error messages:  
        - `the watchdog is stuck with following active activities... `
        - `the watchdog is stuck and the last started / DONE activity is X`
  b) [stretch] We might piggy back on it and collect metrics what percent of time specifica activities are active state